tt_metal/impl/dispatch/kernels/cq_dispatch.cpp

```diff
@@ -713,51 +678,30 @@ void process_write_packed_large() {
-                // Check for block completion
+            // More data needs to be written, but we've exhausted the CB. Acquire more pages.
+            if (data_ptr == cb_fence) {
                 if (cb_fence == block_next_start_addr[rd_block_idx]) {
-                    // No more writes from this block. Decrement the number of writes
-                    // since they were all accounted for.
-                    uint32_t orphan_size = cb_fence - data_ptr;
-                    writes -= (orphan_size == 0);
-                    mcasts -= (orphan_size == 0) * num_dests;
-                    // Check for dispatch_cb wrap
                     if (rd_block_idx == dispatch_cb_blocks - 1) {
-                        ASSERT(cb_fence == dispatch_cb_end);
-                        if (orphan_size != 0) {
-                            cq_noc_async_write_with_state<CQ_NOC_SnDL>(data_ptr, dst_addr, orphan_size, num_dests);
-                            length -= orphan_size;
-                            xfer_size -= orphan_size;
-                            dst_addr += orphan_size;
-                            // All writes from this block have completed.
-                            orphan_size = 0;
-                        }
                         cb_fence = dispatch_cb_base;
                         data_ptr = dispatch_cb_base;
                     }
-
+                    // Block completion - account for all writes issued for this block before moving to next
                     block_noc_writes_to_clear[rd_block_idx] += writes;
                     noc_nonposted_writes_num_issued[noc_index] += writes;
                     writes = 0;
                     move_rd_to_next_block<dispatch_cb_blocks>(block_noc_writes_to_clear, rd_block_idx);
-                    // Next write will be from next block. "Reserve" pages for it.
-                    writes += (orphan_size == 0);
-                    mcasts += (orphan_size == 0) * num_dests;
                 }
-
-                // Wait for dispatcher to supply a page (this won't go beyond the buffer end)
                 uint32_t n_pages = cb_acquire_pages<my_noc_xy, my_dispatch_cb_sem_id, dispatch_cb_log_page_size>(
                     cb_fence, block_next_start_addr, rd_block_idx);
                 cb_fence += n_pages * dispatch_cb_page_size;
             }
-
-            cq_noc_async_write_with_state<CQ_NOC_SnDL>(data_ptr, dst_addr, xfer_size, num_dests);
-
+            // Transfer size is min(remaining_length, data_available_in_cb)
+            uint32_t available_data = cb_fence - data_ptr;
+            uint32_t xfer_size = (length > available_data) ? available_data : length;
+            cq_noc_async_write_with_state_any_len(data_ptr, dst_addr, xfer_size, num_dests);
+            uint32_t num_noc_packets_written = (xfer_size + NOC_MAX_BURST_SIZE - 1) / NOC_MAX_BURST_SIZE;
+            writes += num_noc_packets_written;
+            mcasts += num_dests * num_noc_packets_written;
```

Comment:
To minimize the inc/mul for this mcast variable, we can actually just do `mcasts += num_dests * writes;` whenever we are done with the block/about to zero out `writes` var

Reply:
Done